### PR TITLE
@safe destruction for Unique

### PR DESCRIPTION
**Pull request by dukc**
Destroying Unique class has been made @safe, but constructing them, except from
other Unique objects, is @system because that's where the error happened if there
remain dangling references.

The destructor inspects if it's actually safe to destroy the content at end of scope.

Also refactored the type a bit.

### CODE REVIEW

File: std/typecons.d [hunk 1]
```diff
@@ -52,6 +52,9 @@ deleted at the end of the scope, unless it is transferred.  The
 transfer can be explicit, by calling $(D release), or implicit, when
 returning Unique from a function. The resource can be a polymorphic
 class object, in which case Unique behaves polymorphically too.
+It is @safe to use as long as it's resource is, but constructing it is
```
Comment (wilzbach) [inline std/typecons.d]: All D keywords should be put in backticks

File: std/typecons.d [hunk 2]
```diff
@@ -143,25 +146,39 @@ public:
         u._p = null;
     }
 
-    ~this()
+    // whether a RefT instance can be @safely destroyed at end of scope
+    static if (__traits(compiles, (RefT t) @safe {})) @trusted ~this()
```
Comment (wilzbach) [inline std/typecons.d]: Please put this inside the method as such implementation details shouldn't affect the user and as ddoc doesn't know about `static if`

File: std/typecons.d [hunk 3]
```diff
@@ -143,25 +146,39 @@ public:
...
-        if (_p !is null) destroy(_p);
-        _p = null;
+        debug(Unique) writeln("Unique @trusted destructor of ", (_p is null)? null: _p);
+        if (_p !is null)
+        {
+            destroy(_p);
+            _p = null;
+        }
     }
+    else this()
```
Comment (wilzbach) [inline std/typecons.d]: should be `~this`

File: std/typecons.d [hunk 4]
```diff
@@ -143,25 +146,39 @@ public:
...
+    {
+        debug(Unique) writeln("Unique @system destructor of ", (_p is null)? null: _p);
```
Comment (wilzbach) [inline std/typecons.d]: This isn't covered by your test ..

File: std/typecons.d [hunk 5]
```diff
@@ -143,25 +146,39 @@ public:
...
+    */
     Unique release()
     {
-        debug(Unique) writeln("Release");
-        auto u = Unique(_p);
-        assert(_p is null);
-        debug(Unique) writeln("return from Release");
-        return u;
+        debug(Unique) writeln("Unique Release");
+        import std.algorithm : move;
```
Comment (wilzbach) [inline std/typecons.d]: Be more specific with the imports as then the compiler might not need to open and parse the `std.algorithm` package, but only the subpackage (that's why CircleCi fails), here it's `std.algorithm.mutation`

File: std/typecons.d (hunk 3 above)
Comment (dukc) [inline std/typecons.d]: Thanks, you just saved us from a regression. That was a bad mistake from me.

**Comment by dukc** [comment]
After pulling the latest master branch I started to get these errors when compiling:
std\typecons.d(5139): Error: module core.memory import 'pureMalloc' not found
std\typecons.d(5376): Error: template instance std.typecons.RefCounted!(int, cast(RefCountedAutoInitialize)1) error instantiating

I suppose that means I have to download and compile the latest DRuntime before proceeding?

**Comment by jmdavis** [comment]
> I suppose that means I have to download and compile the latest DRuntime before proceeding?

Likely. I always use the latest dmd, druntime, and Phobos together. Any time that they're not in sync, it risks compilation failures.

**Comment by dukc** [comment]
I managed to compile dmd 74 and set it as default, but druntime seems more mysterious. It builds a druntime.lib file, but where I should put it? I can't find any druntime.lib file to replace!

**Comment by wilzbach** [comment]
> I managed to compile dmd 74 and set it as default, but druntime seems more mysterious. It builds a druntime.lib file, but where I should put it? I can't find any druntime.lib file to replace!

Druntime is bundled with Phobos (for releases), but if you follow the D layout and have cloned the repositories in folders of the same parent directory, you can simply execute the newly built dmd executable from the dmd source directory. The included local config file automatically sets the imports for druntime - also most Makefiles expect this layout anyways. If you need to do it manually, have a look at the Makefiles for the guidance to set `-defaultlib` etc.

**Comment by dukc** [comment]
From the source directory... thanks! I had just copied the binary to replace my regular dmd. But what's the local config file? I can't find sc.ini in that place.

**Comment by wilzbach** [comment]
> But what's the local config file? I can't find sc.ini in that place.

It's automatically generated on Linux:

https://github.com/dlang/dmd/blob/master/src/posix.mak#L358

However, it seems that this isn't the case for the Windows Makefiles. I don't use Windows either, but you could create a `sc.ini` that is similar to the Posix one (or pass the mentioned flags directly). In any case you can find the configs used for release builds in `ini` of dmd's root directory.

**Comment by dukc** [comment]
It does. I thought about that too, but didn't bother to mention as long as I can't test my code. I will revert that but retain that release as move thing. Gotta rename the pr then since it's main point has changed.

**Comment by dukc** [comment]
Closed until I can push things mentioned above. I do not want somebody to accidently review this as positive!

**Comment by dukc** [comment]
reopened as another pr: https://github.com/dlang/phobos/pull/5174

**Pull request closed without merging**